File: PRACTICA_JAVA/src/irrgarten/Game.java
package irrgarten;

import java.util.ArrayList;

/**
 * Main game controller class that manages the game state and flow.
 * Handles player turns, combat, movement, and game state updates.
 * 
 * @author marcosbslinux
 */
public class Game {

    /** Maximum number of rounds allowed in a combat */
    private static final int MAX_ROUNDS = 10;

    /** Index of the current player in the players list */
    private int currentPlayerIndex;
    
    /** Game log containing messages about game events */
    private String log;
    
    /** List of players in the game */
    private ArrayList<Player> players;
    
    /** List of monsters in the game */
    private ArrayList<Monster> monsters;
    
    /** Reference to the current player */
    private Player currentPlayer;
    
    /** Game labyrinth containing players and monsters */
    private Labyrinth labyrinth;

    /**
     * Initializes a new game with the specified number of players
     * @param nplayers Number of players in the game
     */
    public Game(int nplayers) {
        log = "";

        //Inicializo players
        players = new ArrayList<Player>();

        //Creo los players y los meto en el contenedor players
        for (int i = 0; i < nplayers; ++i) {
            char c = (char) ('0' + i);
            players.add(new Player(c, Dice.randomIntelligence(), Dice.randomStrength()));
        }
        
        labyrinth = new Labyrinth(10, 10, 3, 3);


        //Esparcir a los jugadores por el tablero
        labyrinth.spreadPlayers(players);

        //Elijo quién empieza la partida
        currentPlayerIndex = Dice.whoStarts(nplayers);

        //Fijo el jugador actual
        currentPlayer = players.get(currentPlayerIndex);

        //Inicializo monsters
        monsters = new ArrayList<Monster>();
        
        configureLabyrinth();

    }

    /**
     * Checks if the game has ended (a player has won)
     * @return true if the game has ended, false otherwise
     */
    public boolean finished() {
        return labyrinth.haveAWinner();
    }

    /**
     * Advances the game by one step, handling player movement and combat
     * @param preferredDirection The player's desired movement direction
     * @return true if the game has ended after this step, false otherwise
     */
    public boolean nextStep(Directions preferredDirection) {
        log = "";
        boolean dead = currentPlayer.dead();
        boolean endGame;
        Directions direction;
        Monster monster;
        GameCharacter winner;

        if (!dead) {
            direction = actualDirection(preferredDirection);

            if (direction != preferredDirection) {
                logPlayerNoOrders();
            }

            monster = labyrinth.putPlayer(direction, currentPlayer);

            if (monster == null) {
                logNoMonster();
            } else {
                winner = combat(monster);
                manageReward(winner);
            }
        } else {
            manageResurrection();
        }

        endGame = finished();

        if (!endGame) {
            nextPlayer();
        }

        return endGame;
    }

    /**
     * Gets the current state of the game
     * @return GameState object containing all relevant game information
     */
    public GameState getGameState() {
        StringBuilder players_toS = new StringBuilder();
        for(int i=0; i<players.size(); ++i)
        {
            players_toS.append(players.get(i).toString());
            players_toS.append("\n");
        }
        
        StringBuilder monsters_toS = new StringBuilder();
        for(int i=0; i<monsters.size(); ++i)
        {
            monsters_toS.append(monsters.get(i).toString());
            monsters_toS.append(", ");
        }
        //CAMBIAR
        GameState state = new GameState(labyrinth.toString(), players_toS.toString(),
                monsters_toS.toString(), currentPlayerIndex,
                finished(), log);
        return state;
    }

    /**
     * Configures the labyrinth by placing monsters in their positions
     */
    private void configureLabyrinth() {
        // Set labyrinth dimensions and number of monsters
         final int rows = 10;
         final int cols = 10;

         // Add outer walls to the labyrinth
         labyrinth.addBlock(Orientation.HORIZONTAL, 0, 0, cols);
         labyrinth.addBlock(Orientation.HORIZONTAL, rows-1, 0, cols);
         labyrinth.addBlock(Orientation.VERTICAL, 1, 0, rows);
         labyrinth.addBlock(Orientation.VERTICAL, 1, cols-1, rows);

         // Create and add monsters to the labyrinth
         Monster monster = new Monster("1", 200, 200);
         labyrinth.addMonster(2, 6, monster);

         Monster monster2 = new Monster("1", Dice.randomIntelligence(), Dice.randomStrength());
         labyrinth.addMonster(7, 4, monster2);
         
         monsters.add(monster);
         monsters.add(monster2);
        
    }

    /**
     * Advances to the next player's turn
     */
    private void nextPlayer() {
        //Pasar al siguiente jugador
        currentPlayerIndex = (currentPlayerIndex + 1) % players.size();
        currentPlayer = players.get(currentPlayerIndex);
    }

    /**
     * Determines the actual movement direction based on player input and valid moves
     * @param preferredDirection The player's desired movement direction
     * @return The actual direction the player will move
     */
    private Directions actualDirection(Directions preferredDirection) {
        int currentRow = currentPlayer.getRow();
        int currentCol = currentPlayer.getCol();

        ArrayList<Directions> validMoves = labyrinth.validMoves(currentRow, currentCol);

        Directions output = currentPlayer.move(preferredDirection, validMoves);

        return output;
    }

    /**
     * Handles combat between the current player and a monster
     * @param monster The monster to fight against
     * @return The winner of the combat (PLAYER or MONSTER)
     */
    private GameCharacter combat(Monster monster) {
        int rounds = 0;
        GameCharacter winner = GameCharacter.PLAYER;
        float playerAttack = currentPlayer.attack();
        float monsterAttack;
        boolean lose = monster.defend(playerAttack);

        while (!lose && rounds < MAX_ROUNDS) {
            winner = GameCharacter.MONSTER;
            rounds++;

            monsterAttack = monster.attack();
            lose = currentPlayer.defend(monsterAttack);

            if (!lose) {
                playerAttack = currentPlayer.attack();
                winner = GameCharacter.PLAYER;
                lose = monster.defend(playerAttack);
            }
        }
        logRounds(rounds, MAX_ROUNDS);

        return winner;
    }

    /**
     * Manages rewards after combat based on the winner
     * @param winner The winner of the combat (PLAYER or MONSTER)
     */
    private void manageReward(GameCharacter winner) {
        if (winner == GameCharacter.PLAYER) {
            currentPlayer.receiveReward();
            logPlayerWon();
        } else {
            logMonsterWon();
        }
    }

    /**
     * Handles player resurrection logic
     */
    private void manageResurrection() {
        boolean resurrect = Dice.resurrectPlayer();

        if (resurrect) {
            currentPlayer.resurrect();
            logResurrected();
        } else {
            logPlayerSkipTurn();
        }
    }

    /**
     * Logs a message when the player wins combat
     */
    private void logPlayerWon() {
        log += "El jugador " + currentPlayer + " ha ganado el combate.\n";
    }

    /**
     * Logs a message when the monster wins combat
     */
    private void logMonsterWon() {
        log += "El monstruo ha ganado el combate.\n";
    }

    /**
     * Logs a message when a player is resurrected
     */
    private void logResurrected() {
        log += "El jugador " + currentPlayer + " ha resucitado.\n";
    }

    /**
     * Logs a message when a player skips their turn due to being dead
     */
    private void logPlayerSkipTurn() {
        log += "El jugador " + currentPlayer + " ha perdido el turno por estar muerto.\n";
    }

    /**
     * Logs a message when a player cannot follow movement orders
     */
    private void logPlayerNoOrders() {
        log += "El jugador " + currentPlayer + " no ha podido seguir las instrucciones del jugador humano.\n";
    }

    /**
     * Logs a message when a player moves to an empty cell
     */
    private void logNoMonster() {
        log += "El jugador " + currentPlayer + " se ha movido a una celda vacía o no le ha sido posible moverse.\n";
    }

    /**
     * Logs the number of rounds played in combat
     * @param rounds Number of rounds played
     * @param max Maximum number of rounds allowed
     */
    private void logRounds(int rounds, int max) {
        log += "Se han jugado " + Integer.toString(rounds) + " de " + Integer.toString(max) + " rondas.\n";
    }
}

File: PRACTICA_RUBY/Game.rb
module Irrgarten
  # Main game controller class that manages game state and flow
  class Game
    # Maximum number of combat rounds
    @@MAX_ROUNDS=10

    # Initializes a new game with specified number of players
    # @param nplayers [Integer] number of players in the game
    def initialize(nplayers)
      @current_player_index=Dice.who_starts(nplayers)

      @players = Array.new(nplayers) 

      #Inicializa con players
      for i in 0...nplayers
        @players << Player.new(i.chr, Dice.random_intelligence, Dice.random_strength)
      end
      @current_player=players[@current_player_index]
      @labyrinth=Labyrinth.new
      @labyrinth.spread_players(@players)

      current_player
      @monsters=Array.new
      @log = ""
    end

    # Checks if game has finished (a player reached the exit)
    # @return [Boolean] true if game is finished
    def finished
      @labyrinth.have_a_winner
    end

    # Advances game state by one step
    # @param preferred_direction [Symbol] player's desired movement direction
    # @return [Boolean] true if game has finished after this step
    def next_step(preferred_direction)
      @log = ""
      dead = @current_player.dead

      if !dead
        direction = actual_direction(preferred_direction)

        if direction != preferred_direction
          log_player_no_orders
        end

        monster = @labyrinth.put_player(direction, @current_player)

        if monster == nil
          log_no_monster
        else
            winner = combat(monster)
            manage_reward(winner)
        end
      else
        manage_resurrection
      end

      end_game = finished
      
      if !end_game
        next_player
      end
      end_game
    end

    # Gets current game state
    # @return [GameState] object containing game state information
    def get_game_state
      players_s = String.new
      monsters_s = String.new

      for i in 0...@players.size
        players_s << @players[i].to_s
        players_s << ", "
      end

      for i in 0...@monsters.size
        monsters_s << @monsters[i].to_s
        monsters_s << ", "
      end 
      Gamestate.new(@labyrinth.to_s, players_s, monsters_s, current_player_index, finished, log)
    end
    
    private

    # Configures labyrinth (implementation pending)
    def configure_labyrinth
      #Aún no sé cómo se hace
    end

    # Advances to next player in turn order
    # @return [nil]
    def next_player
      @current_player_index = (@current_player_index + 1) % @players.size
      @current_player = @players[@current_player_index]
    end

    # Determines actual movement direction considering valid moves
    # @param preferred_direction [Symbol] player's desired direction
    # @return [Symbol] actual movement direction
    def actual_direction(preferred_direction)
      current_row = @current_player.row
      current_col = @current_player.col
      
      valid_moves = @labyrinth.valid_moves(current_row, current_col)

      output = @current_player.move(preferred_direction, valid_moves)

      output
    end

    # Handles combat between player and monster
    # @param monster [Monster] monster to fight
    # @return [Symbol] winner of combat (GameCharacter::PLAYER or GameCharacter::MONSTER)
    def combat(monster)      
      rounds = 0
      winner = GameCharacter::PLAYER
      player_attack = @curent_player.attack
      lose = monster.defend(player_attack)

      while !lose && rounds < @@MAX_ROUNDS
        winner = GameCharacter::MONSTER
        rounds+=1

        monster_attack = monster.attack
        lose = @current_player.defend(monster_attack)

        if !lose
          player_attack = @current_player.attack
          winner = GameCharacter::PLAYER
          lose = monster.defend(player_attack)
        end
      end
      log_rounds(rounds, @@MAX_ROUNDS)
      winner
    end

    # Manages rewards after combat
    # @param winner [Symbol] winner of combat
    # @return [nil]
    def manage_reward(winner)
      if winner == GameCharacter::PLAYER
        @current_player.receive_reward
        log_player_won
      else
        log_monster_won
      end
      nil
    end

    # Handles player resurrection logic
    # @return [nil]
    def manage_resurrection
      resurrect = Dice.resurrect_player
      
      if resurrect
        @current_player.resurrect
      else
        log_player_skip_turn
      end
      nil
    end

    # Logs player victory message
    # @return [nil]
    def log_player_won
      @log+="El jugador #{@current_player.to_s} ha ganado el combate \n"
    end

    # Logs monster victory message
    # @return [nil]
    def log_monster_won
      @log+="El jugador monstruo ha ganado el combate \n"
    end

    # Logs player resurrection message
    # @return [nil]
    def log_resurrected
      @log+="El jugador #{@current_player.to_s} ha resucitado \n"
    end

    # Logs player skip turn message
    # @return [nil]
    def log_player_skip_turn
      @log+="El jugador #{@current_player.to_s} ha perdido el turno por estar muerto \n"
    end

    # Logs player unable to follow orders message
    # @return [nil]
    def log_player_no_orders
      @log+="El jugador #{@current_player.to_s} no ha podido seguir las instrucciones del jugador humano \n"
    end

    # Logs no monster encounter message
    # @return [nil]
    def log_no_monster
      @log+="El jugador #{@current_player.to_s} se ha movido a una celda vacía o no le ha sido posible moverse \n"
    end

    # Logs combat rounds information
    # @param rounds [Integer] number of rounds fought
    # @param max [Integer] maximum possible rounds
    # @return [nil]
    def log_rounds(rounds, max)
      @log+="Se han jugado #{rounds} de #{max} rondas \n"
    end
  end
end
